RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java

```diff
@@ -262,12 +264,76 @@ public static RoaringBitmap naive_and(RoaringBitmap... bitmaps) {
       return new RoaringBitmap();
     }
     RoaringBitmap answer = bitmaps[0].clone();
-    for (int k = 1; k < bitmaps.length; ++k) {
+    for (int k = 1; k < bitmaps.length && !answer.isEmpty(); ++k) {
       answer.and(bitmaps[k]);
     }
     return answer;
   }
 
+  /**
+   * Computes the intersection by first intersecting the keys, avoids
+   * materialising containers.
+   *
+   * @param bitmaps the inputs
+   * @return the intersection of the bitmaps
+   */
+  public static RoaringBitmap workShyAnd(RoaringBitmap... bitmaps) {
+    long[] words = new long[1024];
+    RoaringBitmap first = bitmaps[0];
+    for (int i = 0; i < first.highLowContainer.size; ++i) {
+      char key = first.highLowContainer.keys[i];
+      words[key >>> 6] |= 1L << key;
+    }
+    int numContainers = first.highLowContainer.size;
+    for (int i = 1; i < bitmaps.length && numContainers > 0; ++i) {
+      numContainers = Util.intersectArrayIntoBitmap(words,
```

Comment:
what is the purpose of this loop, looks like numContainers gets overridden.

Reply:
Yes, it is overridden but once it reaches zero, the loop stops.

Reply:
That’s the idea, this is a running intersection count, “words” contains the intersection so far, and we compute the bit count as we do the intersection between words and keys, and get out as soon as the intersection is empty. It looks weird but is basically sane.